index.bs

```diff
@@ -815,21 +820,113 @@ WebXR provides various transforms in the form of <dfn lt="matrix">matrices</dfn>
+};
+</pre>
+
+<div class="algorithm" data-algorithm="construct-rigid-transform">
+
+The <dfn constructor for="XRRigidTransform">XRRigidTransform(|position|, |orientation|)</dfn> constructor MUST perform the following steps when invoked:
+
+  1. Let |transform| be a new {{XRRigidTransform}}.
+  1. If |position| is not a {{DOMPointInit}} initialize |transform|'s {{XRRigidTransform/position}} to <code>{ x: 0.0, y: 0.0, z: 0.0, w: 1.0 }</code>.
+  1. Else initialize |transform|'s {{XRRigidTransform/position}}’s {{DOMPointReadOnly/x}} value to |position|'s x dictionary member, {{DOMPointReadOnly/y}} value to |position|'s y dictionary member, {{DOMPointReadOnly/z}} value to |position|'s z dictionary member and {{DOMPointReadOnly/w}} to <code>1.0</code>.
+  1. If |orientation| is not a {{DOMPointInit}} initialize |transform|'s {{XRRigidTransform/orientation}} to <code>{ x: 0.0, y: 0.0, z: 0.0, w: 1.0 }</code>.
+  1. Else initialize |transform|'s {{XRRigidTransform/orientation}}’s {{DOMPointReadOnly/x}} value to |orientation|'s x dictionary member, {{DOMPointReadOnly/y}} value to |orientation|'s y dictionary member, {{DOMPointReadOnly/z}} value to |orientation|'s z dictionary member and {{DOMPointReadOnly/w}} value to |orientation|'s w dictionary member.
+  1. [=Normalize=] |transform|'s {{XRRigidTransform/orientation}}.
+  1. Return |transform|.
+
+</div>
+
+The <dfn attribute for="XRRigidTransform">position</dfn> attribute is a 3-dimensional point, given in meters, describing the translation component of the transform. The {{XRRigidTransform/position}}'s {{DOMPointReadOnly/w}} attribute MUST be <code>1.0</code>.
+
+The <dfn attribute for="XRRigidTransform">orientation</dfn> attribute is a quaternion describing the rotational component of the transform. The {{XRRigidTransform/orientation}} MUST be normalized to have a length of <code>1.0</code>.
+
+The <dfn attribute for="XRRigidTransform">matrix</dfn> attribute returns the transform described by the {{XRRigidTransform/position}} and {{XRRigidTransform/orientation}} attributes as a [=matrix=].
+
+ISSUE: Define <dfn>Normalize</dfn>
+
+XRRay {#xrray-interface}
+-----
+
+An {{XRRay}} is a geometric ray described by a {{XRRay/origin}} point and {{XRRay/direction}} vector.
+
+<pre class="idl">
+[SecureContext, Exposed=Window,
+ Constructor(optional DOMPointInit origin, optional DOMPointInit direction),
+ Constructor(XRRigidTransform transform)]
+interface XRRay {
+  readonly attribute DOMPointReadOnly origin;
+  readonly attribute DOMPointReadOnly direction;
+  readonly attribute Float32Array matrix;
+};
+</pre>
+
+<div class="algorithm" data-algorithm="construct-ray-origin-direction">
+
+The <dfn constructor for="XRRay">XRRay(|origin|, |direction|)</dfn> constructor MUST perform the following steps when invoked:
+
+  1. Let |ray| be a new {{XRRay}}.
+  1. If |origin| is not a {{DOMPointInit}} initialize |ray|'s {{XRRay/origin}} to <code>{ x: 0.0, y: 0.0, z: 0.0, w: 1.0 }</code>.
```

Comment:
Actually... thinking on this again, should it be:
> 1. If |origin| is <code>null</code> initialize |ray|'s {{XRRay/origin}} to <code>{ x: 0.0, y: 0.0, z: 0.0, w: 1.0 }</code>.
> 1. If |origin| is not a {{DOMPointInit}} initialize throw an {{InvalidStateError}}

And same for {{XRRigidTransform}}?

Reply:
I will double check, but to know knowledge when you specify a parameter as optional it actually comes through as `undefined` instead of `null`, which is why I phrased it the way that I did (because I would want `undefined` and `null` to both cause an init to the default values.)

That says, your point still stands I think. It ought to be whatever the spec-language equivalent of:
```
if (!origin) { // default }
else if (origin instanceOf DOMPointInit) { // Set value }
else { throw }
```

But event then, now that I"m writing it out like that, `DOMPointInit` is actually a dictionary, so doing "instance of" doesn't really work. :( Lemme think about this some more.